platform/x11/cog-platform-x11.c

```diff
@@ -0,0 +1,896 @@
+            if (configure_notify->width == xcb_data.width
+                && configure_notify->height == xcb_data.height)
+                break;
+
+            xcb_data.width = configure_notify->width;
+            xcb_data.height = configure_notify->height;
+
+            wpe_view_backend_dispatch_set_size (wpe_view_data.backend,
+                                                xcb_data.width,
+                                                xcb_data.height);
+            xcb_schedule_repaint ();
+            break;
+        }
+        case XCB_CLIENT_MESSAGE:
+        {
+            xcb_client_message_event_t *client_message = (xcb_client_message_event_t *) event;
+            if (client_message->window != xcb_data.window)
+                break;
+
+            if (client_message->type == xcb_data.atom_wm_protocols
+                && client_message->data.data32[0] == xcb_data.atom_wm_delete_window) {
+                g_application_quit (g_application_get_default ());
+                break;
+            }
+
+            if (client_message->type == XCB_ATOM_NOTICE) {
+                if (xcb_data.needs_initial_paint) {
+                    xcb_initial_paint ();
+                    xcb_data.needs_initial_paint = false;
+                }
+
+                if (xcb_data.needs_frame_completion) {
+                    xcb_frame_completion ();
+                    xcb_data.needs_frame_completion = false;
+                }
+                break;
+            }
+            break;
+        }
+        case XCB_KEY_PRESS:
+            xcb_handle_key_press ((xcb_key_press_event_t *) event);
+            break;
+        case XCB_KEY_RELEASE:
+            xcb_handle_key_release ((xcb_key_release_event_t *) event);
+            break;
+        case XCB_BUTTON_PRESS:
+            xcb_handle_button_press ((xcb_button_press_event_t *) event);
+            break;
+        case XCB_BUTTON_RELEASE:
+            xcb_handle_button_release ((xcb_button_release_event_t *) event);
+            break;
+        case XCB_MOTION_NOTIFY:
+            xcb_handle_motion_event ((xcb_motion_notify_event_t *) event);
+            break;
+        default:
+            break;
+        }
+    }
+};
+
+struct xcb_source {
+    GSource source;
+    GPollFD pfd;
+};
+
+static gboolean
+xcb_source_check (GSource *base)
+{
+    struct xcb_source *source = (struct xcb_source *) base;
+    return !!source->pfd.revents;
+}
+
+static gboolean
+xcb_source_dispatch (GSource *base, GSourceFunc callback, gpointer user_data)
+{
+    struct xcb_source *source = (struct xcb_source *) base;
+    if (source->pfd.revents & (G_IO_ERR | G_IO_HUP))
+        return FALSE;
+
+    xcb_process_events ();
+    source->pfd.revents = 0;
+    return TRUE;
+}
+
+static xcb_atom_t
+get_atom (struct xcb_connection_t *connection, const char *name)
+{
+    xcb_intern_atom_cookie_t cookie;
+    xcb_intern_atom_reply_t *reply;
+    xcb_atom_t atom;
```

Comment:
We are building in C99 mode, these variables could be assigned as soon as they are defined, and there is no need to define them all at the beginning of the function. Also we only support Unixish systems, so using `g_autofree`, `g_autoptr()`, and friends is not only allowed, but encouraged.